sympy/stats/stochastic_process_types.py

```diff
@@ -615,6 +662,36 @@ class DiscreteMarkovChain(DiscreteTimeStochasticProcess, MarkovProcess):
     >>> P(Eq(Y[3], 2), Eq(Y[1], 1)).round(2)
     0.36
 
+    Probabilities will be calculated based on indexes rather
+    than state names. For example, with the Sunny-Cloudy-Rainy
+    model with string state names:
+
+    >>> Y = DiscreteMarkovChain("Y", ['Sunny', 'Cloudy', 'Rainy'], T)
+    >>> P(Eq(Y[3], 2), Eq(Y[1], 1)).round(2)
+    0.36
```

Comment:
What is the benefit of allowing string state spaces if we're just treating them as integers?

Reply:
For ease of use. If they want to, the user can forget about index values by using `index_of` in probablility statements as the next example suggests.

Maybe one day we could have `P(Eq(Y[2], 'Sunny'))` but I think that would be for another PR.

Reply:
How would you compute expectation values with string values?

Reply:
They currently work using the index of each state. I have no clue how it would work if it was able to work with state names. Maybe using the modal state. Expectations of being in a given state is not that big of a subject in Markov Chains though.

Reply:
I have seen that Bernoulli had similar problem, and it had resolved it through using named Symbols instead of strings though.

Reply:
Here is an idea. May be we can create a class `State` for representing states of stochastic processes. Now the objects of this class or it's child classes should be used as states. This way processes won't be concerned about the contents of the state. We can give numerical id starting from 0 to each new state created and this id can used to index the transition matrix, avoiding the need of mapping each state to a number manually. The issue of allowing strings or not will be then upto the user. A probable interface can be for class `State` can be,

```python
class State(Base):

    id = 0
    
    def __new__(cls, contents=None):
        ...

    def __eq__(self, other):
        return self.contents == other.contents

    # other operators can be defined accordingly. 
```
Thoughts?

Reply:
Although a new class to deal with bijective relation between string valued states and integer valued states can be reasonable, I think that a more lightweight solution can be making named states as symbols,
such that we can compute expectation values like `sunny/2 + cloudy/3 + rainy/6`.
I think that it makes sense to substitute `sunny` with `0`, `cloudy` with `1`, ... to get expectation values as desired.

Reply:
I think `Str` state spaces may be better since this allows for states such as `'0 kilometres per hour'`, `'10 kilometres per hour'` or `''`. This allows for far greater flexibility at no cost. Your suggested version, using `Bernoulli` as inspiration, also has its benefits but nothing is stopping the user from doing the following instead:

```python
from sympy import *
from sympy.stats import *

scr = symbols('sunny cloudy rainy')
sunny = scr[0]
X = DiscreteMarkovChain('X', scr)

expectation = sum([state*P(Eq(X[1], X.index_of[state]), Eq(X[0], X.index_of[sunny])) 
                   for state in scr])
print(expectation)
```

This gives the same output as you desire. That being `cloudy*_T[0, 1] + rainy*_T[0, 2] + sunny*_T[0, 0]`. 

As can be seen from this output, `Bernoulli` does not need a matrix. But Markov Chains need it and there is no way to index them based on symbols or strings. This is why I think it should be clear to the user that it does work directly on the state index rather than state names. This should avoid a whole lot of confusion.

Reply:
You can still create symbol instances with `0 kilometres per hour` like `Symbol('0 kilometres per hour')` or `Symbol(r'\text{0 kilometres per hour}')`.
Although it can be weird, I think that using symbol-valued expectation is the best fallback method we can give for any users who try to adventure too much into computing the expectation values for markov chains.
And I'd anticipate little cost for switching Str to Symbol in the middle of development.

The problem of indexing the matrix is just a problem of using `Str` indices before we introduce any good way to depict such bijective mapping between strings and integers.

Unlike @czgdp1807 's idea, I'd want to see a module that can be used in other parts of sympy than only restricted to stats module.
Or other objects may be sufficient for this. For example, a `Tuple` can be used in place of `Set` state space, and it can also be a good representation of a bijective map, and we can use `.index` to look up for the integer valued index.

Reply:
> You can still create symbol instances with 0 kilometres per hour like Symbol('0 kilometres per hour') or Symbol(r'\text{0 kilometres per hour}').

I tried it with `Bernoulli` and it threw an error so I thought it had the same effect for `Symbol`. My mistake.

To reach your goal, I would need to redesign a bit of the working of `expectation` for Markov Chains. I was really hoping to avoid `probability` and `expectation` as much as possible. If I can't get it to work first try, I'll leave it for another PR.

I the state_space-state_index relationship is not that big of a deal as to include a class. If `CommunicationClass` was not needed, I do not think `State` would be needed either.

Reply:
Ok, it was not difficult but I would really prefer not to work on it again. I thought this PR would have gone through a week ago. I still have about 3 more planned PRs that all build on one another. 💀